### PR TITLE
style: Support anonymous box pseudo-elements

### PR DESCRIPTION
**Pull request by emilio**
This is a work-in-progress that:
- Adds support for some pseudo-elements to skip the cascade entirely, in an analogous way to Gecko's anonymous box pseudo-elements.
- Takes rid of `StylistWrapper`, and uses `Arc::get_mut` instead.
- Uses the first bullet to precompute the `-servo-details-content` pseudo's style.

I'd like @bholley to take a look before following, do you think that the aproach is the correct?
Also, @SimonSapin could want to put some eyes on it.

Depends on https://github.com/servo/rust-selectors/pull/81

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/10815)

### CODE REVIEW

**Comment by bholley** [comment]
Nice! A few random questions after a very brief and incomplete look at this:
* What changed so that we can now get rid of StylistWrapper so easily?
* The selectors for -servo-details-content in the UA stylesheet are not of the form *|*. Presumably that means they can't use this system, right? Should we change the UA stylesheet?
* If the data is the same for the whole document, hanging a separate arc to it off of every node seems suboptimal. Is that what's happening here, or did I miss something?

**Comment by emilio** [comment]
> * What changed so that we can now get rid of StylistWrapper so easily?

Nothing, `Arc::get_mut` I guess, but this could have been changed a long
time before if I'm not wrong.

> * The selectors for -servo-details-content in the UA stylesheet are
> not of the form *|*. Presumably that means they can't use this system,
> right? Should we change the UA stylesheet?

Yes, in fact I changed them (see [the
diff](https://github.com/servo/servo/pull/10815/commits/9fb977db47077305ab0975472cdba9b3f5a1bc65#diff-f3099215c0563b03e07e5529c6edb190R54))

> * If the data is the same for the whole document, hanging a separate
> arc to it off of every node seems suboptimal. Is that what's happening
> here, or did I miss something?

Agreed, but we can't do anything else other than passing around the
precomputed data as an argument everywhere else, and that's a huge
change I didn't felt like making for this PR (due to time constraints).

The cool part is that with this change we can pass over more common data
to the nodes without more overhead, but agreed that a common structure
passed from the callers would be more optimal.

Now I think more about it, iff layout workers are per-document (which
right now I think they are), and `initialize_data` is called from that
same thread (I *think* it is, I'd have to look at the code), we could
rely on a thread-local store updated just when we call `Stylist::update`
from the layout thread, without having to do huge changes to layout
traversal.

That'd reduce the overhead from one Arc clone per node to one clone per
layout worker. What do you think about that? To get zero cloning you'll
have to lock when accessing it if I'm not wrong.

If you think this can't land without it, let me know and I'll make one
of those changes (though the first one is pretty big and prone to bitrot
for my limited time right now).

**Comment by emilio** [comment]
Sorry for the extra comment, but Github ate the quotes on my previous one, and doesn't want to show them again.

> What changed so that we can now get rid of StylistWrapper so easily?

Nothing, `Arc::get_mut` I guess, but this could have been changed a long
time before if I'm not wrong.

> The selectors for -servo-details-content in the UA stylesheet are
> not of the form *|*. Presumably that means they can't use this system,
> right? Should we change the UA stylesheet?

Yes, in fact I changed them (see [the
diff](https://github.com/servo/servo/pull/10815/commits/9fb977db47077305ab0975472cdba9b3f5a1bc65#diff-f3099215c0563b03e07e5529c6edb190R54))

> If the data is the same for the whole document, hanging a separate
> arc to it off of every node seems suboptimal. Is that what's happening
> here, or did I miss something?

Agreed, but we can't do anything else other than passing around the
precomputed data as an argument everywhere else, and that's a huge
change I didn't felt like making for this PR (due to time constraints).

The cool part is that with this change we can pass over more common data
to the nodes without more overhead, but agreed that a common structure
passed from the callers would be more optimal.

Now I think more about it, iff layout workers are per-document (which
right now I think they are), and `initialize_data` is called from that
same thread (I *think* it is, I'd have to look at the code), we could
rely on a thread-local store updated just when we call `Stylist::update`
from the layout thread, without having to do huge changes to layout
traversal.

That'd reduce the overhead from one Arc clone per node to one clone per
layout worker. What do you think about that? To get zero cloning you'll
have to lock when accessing it if I'm not wrong.

If you think this can't land without it, let me know and I'll make one
of those changes (though the first one is pretty big and prone to bitrot
for my limited time right now).

**Comment by SimonSapin** [comment]
> That'd reduce the overhead from one Arc clone per node to one clone per layout worker

`Arc::clone` is one atomic integer increment, which sounds sounds relatively cheap compared to everything we do to cascade styles for one node. If it’s easy go for it, but I’d say don’t spend too much time on this.

That said I’m completely guessing, you’d have to benchmark it to be sure.

**Comment by emilio** [comment]
Yeah, I think @bholley's concern is more memory-related (one extra pointer per node), but I feel the same here, I'd rather focus on be correct and later speed things up if required.

**Comment by emilio** [comment]
@bholley: I dind't have thought about inheritance, which Gecko relies a ton on.

Just added a few patches to add support for it.

**Comment by bholley** [comment]
Yeah, I was worried about memory rather than clone overhead. We have memory tests that complain whenever Node gets bigger. The memory reporters don't see into PrivateStyleData so the test doesn't notice, but this does grow Node by one pointer.

In general I think passing some sort of general context from the caller is the right thing to do. Is it just annoying to pass, or does it also require parameterizing a bunch of stuff? It seems like having a SharedStyleContext available at most points in the cascade is probably worthwhile. I'm fine for it to be done in a followup though if it adds too much churn in this PR.

I'd like to avoid relying on TLS too much if possible, since some of the thread invariants may change (i.e. we will probably end up sharing threads between documents at some point).

**Comment by emilio** [comment]
@bholley: Added a bunch of changes that take rid of the extra field, take a look.

Also, building stylo I get a crash in a `Can't handle namespaces yet` assertion.

I'll take a look at it later if I have the time, but it's trivial to implement`Servo_GetComputedValuesForAnonymousBox` if we pass the shared context to it now.

I made a few changes to pseudo-elements, so I'm going to do a try run:

@bors-servo: try

**Comment by emilio** [comment]
@bors-servo: try

It seems that the toml `[replaces]` syntax must be at a top-level target, so I had to add the directive for all the ports.

File: components/layout/wrapper.rs [hunk 1]
```diff
@@ -681,42 +691,36 @@ pub trait ThreadSafeLayoutNode : Clone + Copy + Sized + PartialEq {
...
     fn get_after_pseudo(&self) -> Option<Self> {
-        self.borrow_layout_data().unwrap()
-            .style_data.per_pseudo
-            .get(&PseudoElement::After)
-            .map(|style| {
-                self.with_pseudo(PseudoElementType::After(style.get_box().display))
-            })
+        if self.borrow_layout_data().unwrap()
+               .style_data.per_pseudo
+               .contains_key(&PseudoElement::Before) {
```
Comment (emilio) [inline components/layout/wrapper.rs]: Whoops, what a embarrasing typo, `PseudoElement::After` should be here, there should be no more test failures after that I think.

**Comment by emilio** [comment]
That's better :)

r? @bholley and @SimonSapin

**Comment by bholley** [comment]
Thanks for doing this!

r=me on the stuff in geckolib. I'll let @SimonSapin review the style/layout stuff, but let me know Simon if there are any parts you want me to look at.

---

Reviewed 4 of 25 files at r4.
Review status: 4 of 30 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[ports/geckolib/selector_impl.rs, line 237 \[r6\]](https://reviewable.io:443/reviews/servo/servo/10815#-KGGDxHQ852OI_eQptsF:-KGGDxHRY-ZyoanWZQnB:1625640313) ([raw file](https://github.com/servo/servo/blob/1b7726cfb399926bcce82ef2d1a24ca67ef3c56a/ports/geckolib/selector_impl.rs#L237)):*
Why add FirstLine to eager? Seems like this is a behavior change.

---

*[ports/gonk/Cargo.lock, line 1698 \[r6\]](https://reviewable.io:443/reviews/servo/servo/10815#-KGGD6yu3DAwLpiWnqTC:-KGGD6yvrO_O-V_5q-Pq:1779823954) ([raw file](https://github.com/servo/servo/blob/1b7726cfb399926bcce82ef2d1a24ca67ef3c56a/ports/gonk/Cargo.lock#L1698)):*
Presumably we'll need to land the rust-selectors fix first?

---

*[ports/gonk/Cargo.lock, line 1715 \[r6\]](https://reviewable.io:443/reviews/servo/servo/10815#-KGGDETkd24FXlTF0maG:-KGGDETkd24FXlTF0maH:1662569705) ([raw file](https://github.com/servo/servo/blob/1b7726cfb399926bcce82ef2d1a24ca67ef3c56a/ports/gonk/Cargo.lock#L1715)):*
Same here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10815#-:-KGGGGzlM0nvWDOE7lDO:-1626447604)*

**Comment by emilio** [comment]
Review status: 4 of 30 files reviewed at latest revision, 4 unresolved discussions.

---

*[ports/geckolib/selector_impl.rs, line 237 \[r6\]](https://reviewable.io:443/reviews/servo/servo/10815#-KGGDxHQ852OI_eQptsF:-KGGtehxjF9mSwDtnbng:-1432806081) ([raw file](https://github.com/servo/servo/blob/1b7726cfb399926bcce82ef2d1a24ca67ef3c56a/ports/geckolib/selector_impl.rs#L237)):*
Hum... That's true, I *think* it being eagerly cascade is the correct thing (probably missed to add it in the previous stylo patch). I'll double-check it though. 

---

*[ports/gonk/Cargo.lock, line 1698 \[r6\]](https://reviewable.io:443/reviews/servo/servo/10815#-KGGD6yu3DAwLpiWnqTC:-KGGtWMS7a4eEILrGbAk:-1203858476) ([raw file](https://github.com/servo/servo/blob/1b7726cfb399926bcce82ef2d1a24ca67ef3c56a/ports/gonk/Cargo.lock#L1698)):*
Sure, Simon suggested to use this syntax in order to be able to do a try run before he reviews it, but of course it should land before landing this.

---

*[ports/gonk/Cargo.lock, line 1715 \[r6\]](https://reviewable.io:443/reviews/servo/servo/10815#-KGGDETkd24FXlTF0maG:-KGGtd9rSagn3Wv7Rr-J:1700997001) ([raw file](https://github.com/servo/servo/blob/1b7726cfb399926bcce82ef2d1a24ca67ef3c56a/ports/gonk/Cargo.lock#L1715)):*
Ditto.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10815)*

**Comment by emilio** [comment]
Rebased, commits reordered and tidied up, should be good for finishing the review :)

**Comment by emilio** [comment]
@bholley: The last commit fixes the panic you were seeing in stylo. It was a legit bug, but was masqueraded due to us not unwrapping before.

**Comment by emilio** [comment]
@bholley: Apart from submitting https://github.com/bholley/gecko-dev/pull/6, I added a few commits here to implement the anonymous box stuff and the origin stuff, building on top of it.

**Comment by bholley** [comment]
New geckolib changes look good modulo the ArcHelpers thing.

---

Reviewed 2 of 26 files at r31.
Review status: 1 of 30 files reviewed at latest revision, 6 unresolved discussions.

---

*[ports/geckolib/glue.rs, line 213 \[r56\]](https://reviewable.io:443/reviews/servo/servo/10815#-KGPCGD5ECd80mjR4i7y:-KGPCGD5ECd80mjR4i7z:-2051320653) ([raw file](https://github.com/servo/servo/blob/51152e50921a174d91cf18be3df25e4fe577c42f/ports/geckolib/glue.rs#L213)):*
Long term we should just do this once and stick the result in a hashmap. This is fine for now, I've added a TODO in the etherpad.

---

*[ports/geckolib/glue.rs, line 227 \[r56\]](https://reviewable.io:443/reviews/servo/servo/10815#-KGPCR6d0-oLuw3Yp-FL:-KGPCR6d0-oLuw3Yp-FM:-1508386545) ([raw file](https://github.com/servo/servo/blob/51152e50921a174d91cf18be3df25e4fe577c42f/ports/geckolib/glue.rs#L227)):*
You can use ArcHelpers::with to make this cleaner and avoid the explicit transmute.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10815#-:-KGPCmObmArNsUrGvUz2:983764050)*

**Comment by SimonSapin** [comment]
Reviewed 1 of 25 files at r4, 1 of 30 files at r21, 1 of 26 files at r31, 3 of 25 files at r43, 1 of 11 files at r44, 1 of 4 files at r46, 9 of 15 files at r48, 2 of 3 files at r49, 2 of 2 files at r50, 1 of 1 files at r53, 1 of 1 files at r54, 1 of 1 files at r58.
Review status: 22 of 30 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[components/layout/wrapper.rs, line 803 \[r58\]](https://reviewable.io:443/reviews/servo/servo/10815#-KGW-52_r90nzNf3T0c6:-KGW-52_r90nzNf3T0c7:589801137) ([raw file](https://github.com/servo/servo/blob/c84541e5a604a4ee706de533c6598d362fa4f29c/components/layout/wrapper.rs#L803)):*
No need to change this unless you want to, but `HashMap` also implements the `Index` trait so that `map[key]` is the same as `map.get(key).unwrap()`.

---

*[ports/gonk/Cargo.lock, line 1698 \[r6\]](https://reviewable.io:443/reviews/servo/servo/10815#-KGGD6yu3DAwLpiWnqTC:-KGVbCouOS5T3w_QoTNt:-961824333) ([raw file](https://github.com/servo/servo/blob/1b7726cfb399926bcce82ef2d1a24ca67ef3c56a/ports/gonk/Cargo.lock#L1698)):*
I think you’ll need another `./mach cargo-update -p selectors` to pick up selectors 0.5.5, now that it’s published.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10815)*

**Comment by SimonSapin** [comment]
Looks good to me. I’m OK with merging this assuming the rebase is straight-forward.

@bholley you may want to have another look at geckolib changes, Reviewable shows more of them since your review.

File: ports/geckolib/glue.rs [hunk 2]
```diff
@@ -97,19 +129,40 @@ pub struct ArcHelpers<GeckoType, ServoType> {
...
+        forget(owned);
+
+        result
+    }
+
     pub unsafe fn into(ptr: *mut GeckoType) -> Arc<ServoType> {
         transmute(ptr)
     }
 
+    pub unsafe fn from(owned: Arc<ServoType>) -> *mut GeckoType {
```
Comment (bholley) [inline ports/geckolib/glue.rs]: Nit: I don't see any reason why this needs to be unsafe.

**Comment by bholley** [comment]
geckolib changes look fine. Land away!

File: ports/geckolib/glue.rs (hunk 2 above)
Comment (emilio) [inline ports/geckolib/glue.rs]: Done

**Comment by emilio** [comment]
This fails a bogus unit test as a consequence of https://github.com/servo/rust-selectors/pull/82, I'm fixing it now.

**Pull request merged**